### DISCUSSION TITLE
Compile libMesh failed using HPC Cluster

### DISCUSSION THREAD

**Discussion by PengWei97** [idaholab/moose#21388] [Q&A Getting Started] [answered]
Dear MOOSE experts,

Recently I wanted to link libmesh directly to moose, and completed the installation of PESTs according to[ HPC Cluster](https://mooseframework.inl.gov/getting_started/installation/hpc_install_moose.html), the results are as follows,
![image](https://user-images.githubusercontent.com/56183775/175262007-7a53a9cc-df5e-41c0-b655-07a3e95c08f1.png)

But when compiling libmesh, `./scripts/update_and_rebuild_libmesh.sh`, the following error occurs,

```shell
-- Done configuring core library features ---
---------------------------------------------
---------------------------------------------
----- Configuring for optional packages -----
---------------------------------------------
checking for built-in XDR support... yes
checking for boostlib >= 1.57.0... configure: We could not detect the boost libraries (version 1.57 or higher). If you have a staged boost library (still not installed) please specify $BOOST_ROOT in your environment and do not give a PATH to --with-boost option.  If you are sure you have boost installed, then check your version number looking in <boost/version.hpp>. See http://randspringer.de/boost for more documentation.
<<< External boost installation *not* found... will try to configure for libmesh's internal boost >>>
checking for boostlib >= 1.57.0... yes
<<< Using libmesh-provided boost in ./contrib >>>
checking for boost::movelib::unique_ptr support... yes
note: Checking /lib and /include for MPI
note: Could not find /lib/libmpi(.a/.so/.dylib)
mpicxx -std=gnu++17 Compiler Supports MPI 
<<< Found PETSc 3.16.5 installation in /home/pengwei/projects/libmesh/moose/petsc ... >>>
checking whether we can compile a trivial PETSc program... yes
checking for TAO support via PETSc... yes
configure: error: *** PETSc with Hypre was not found, but --enable-petsc-hypre-required was specified.
Running make -j 1...
make: *** No targets specified and no makefile found.  Stop.

```

Any suggestions or recommendations to fix the problem would be greatly appreciated.

Thank you
Wei

**Comment by milljm** [comment, marked as answer]
MOOSE requires libMesh and PETSc. But also that PETSc to be built with HYPRE support (`--download-hypre=1` during PETSc configure). You can see the options MOOSE requires or recommends in the moose/scripts/configure_petsc.sh script:

```bash
cat moose/scripts/configure_petsc.sh
<trimmed>
  python ./configure --download-hypre=1 \
      --with-shared-libraries=$SHARED \
      "$HDF5_STR" \
      "$HDF5_FORTRAN_STR" \
      "$MAKE_NP_STR" \
      --with-debugging=no \
      --download-fblaslapack=1 \
      --download-metis=1 \
      --download-ptscotch=1 \
      --download-parmetis=1 \
      --download-superlu_dist=1 \
      --download-mumps=1 \
      --download-strumpack=1 \
      --download-scalapack=1 \
      --download-slepc=1 \
      --with-mpi=1 \
      --with-openmp=1 \
      --with-cxx-dialect=C++11 \
      --with-fortran-bindings=0 \
      --with-sowing=0 \
      --with-64-bit-indices \
      "$@"
```

**Reply by milljm** [reply]
Oh! Didn't quite catch you used `./update_and_rebuild_petsc.sh` script.

Can you attach the libmesh_diagnostic.log that was generated? It should be located at `moose/scripts/libmesh_diagnostic.log`

**Reply by PengWei97** [reply]
Yes, I used the `update_and_rebuild_petsc.sh` script to complete the installation of PETSc.

> Can you attach the libmesh_diagnostic.log that was generated? It should be located at moose/scripts/libmesh_diagnostic.log

[libmesh_diagnostic.log](https://github.com/idaholab/moose/files/8967566/libmesh_diagnostic.log)

**Reply by milljm** [reply]
The environment does not seem sane to me at the moment. I see references to two different mamba environments (libmesh and moose):
```pre
PATH=/home/pengwei/mambaforge3/envs/libmesh/bin
                               ^^^^ (libmesh)
CONDA_BACKUP*=/home/pengwei/mambaforge3/envs/moose
                                        ^^^^ (moose)
```
Also, having mambaforge3 installed and active, yet:
```pre
Python:
	/usr/bin/python
	Python 2.7.18
```
is _really_ odd. FYI, after we get PETSc/libMesh sorted, you will need to make Python 3 available. As 2.7 is EOL, and will not work with MOOSE.


Can you perform the following and report back:
```bash
mamba deactivate
mamba deactivate    # again because you can activate one, and install from that one (which might bring in
                    # stuff from the previous creation)
env
mamba activate libmesh  # I am assuming this is what you wanted as this is the last environment you were in?
mamba list              # Curious what is installed, as it looks like you are using /usr/bin/mpich,
                        # but have created/used our Conda install instructions at one point
```
Once I see this, we can try to proceed!

**Reply by PengWei97** [reply]
I changed the default python to python3 and executed `./scripts/update_and_rebuild_libmesh.sh`, the output is as follows,
[libmesh_diagnostic.log](https://github.com/idaholab/moose/files/8968810/libmesh_diagnostic.log)
And the last displayed information is as follows
```shell
-------------------------------------------------------------------------------
Configure complete, now type 'make' and then 'make install'.

---------------------------------------------
--------- Done Configuring libMesh ----------
---------------------------------------------
```

I'm not sure whether libmesh was installed successfully, so I executed `make -j4` in the test path, but the display is as follows,
```shell
(libmesh) pengwei@pengwei-virtual-machine: (next)
~/projects/libmesh/moose/test
$ make -j4
make: python-config: Command not found
make: python-config: Command not found
Compiling C++ (in opt mode) /home/pengwei/projects/libmesh/moose/framework/build/unity_src/geomsearch_Unity.C...
Creating Unity /home/pengwei/projects/libmesh/moose/framework/build/unity_src/splits_Unity.C
Creating Unity /home/pengwei/projects/libmesh/moose/framework/build/unity_src/functions_Unity.C
Creating Unity /home/pengwei/projects/libmesh/moose/framework/build/unity_src/vectorpostprocessors_Unity.C
Creating Unity /home/pengwei/projects/libmesh/moose/framework/build/unity_src/partitioner_Unity.C
Creating Unity /home/pengwei/projects/libmesh/moose/framework/build/unity_src/actions_Unity.C
...
Compiling C++ (in opt mode) /home/pengwei/projects/libmesh/moose/framework/contrib/hit/parse.cc...
Compiling C++ (in opt mode) /home/pengwei/projects/libmesh/moose/framework/contrib/hit/lex.cc...
Compiling C++ (in opt mode) /home/pengwei/projects/libmesh/moose/framework/contrib/hit/braceexpr.cc...
Building and linking /home/pengwei/projects/libmesh/moose/framework/contrib/hit/hit.so...
/home/pengwei/projects/libmesh/moose/framework/contrib/hit/hit.cpp:4:10: fatal error: Python.h: No such file or directory
    4 | #include "Python.h"
      |          ^~~~~~~~~~
compilation terminated.
Compiling C++ (in opt mode) /home/pengwei/projects/libmesh/moose/framework/contrib/exodiff/ED_SystemInterface.C...
Compiling C++ (in opt mode) /home/pengwei/projects/libmesh/moose/framework/contrib/exodiff/create_file.C...
Compiling C++ (in opt mode) /home/pengwei/projects/libmesh/moose/framework/contrib/exodiff/stringx.C...
Compiling C++ (in opt mode) /home/pengwei/projects/libmesh/moose/framework/contrib/exodiff/exo_block.C...
Compiling C++ (in opt mode) /home/pengwei/projects/libmesh/moose/framework/contrib/exodiff/node_set.C...
Compiling C++ (in opt mode) /home/pengwei/projects/libmesh/moose/framework/contrib/exodiff/util.C...
Compiling C++ (in opt mode) /home/pengwei/projects/libmesh/moose/framework/contrib/exodiff/add_to_log.C...
Compiling C++ (in opt mode) /home/pengwei/projects/libmesh/moose/framework/contrib/exodiff/Tolerance.C...
Compiling C++ (in opt mode) /home/pengwei/projects/libmesh/moose/framework/contrib/exodiff/STRINGLIB_tokenize.C...
Compiling C++ (in opt mode) /home/pengwei/projects/libmesh/moose/framework/contrib/exodiff/map.C...
Compiling C++ (in opt mode) /home/pengwei/projects/libmesh/moose/framework/contrib/exodiff/check.C...
Compiling C++ (in opt mode) /home/pengwei/projects/libmesh/moose/framework/contrib/exodiff/SL_tokenize.C...
make: *** [/home/pengwei/projects/libmesh/moose/framework/moose.mk:124: /home/pengwei/projects/libmesh/moose/framework/contrib/hit/hit.so] Error 1
make: *** Waiting for unfinished jobs....

```
This may also be a problem with the python settings.

> Can you perform the following and report back: ...

It is displayed as follows,
```shell
(libmesh) pengwei@pengwei-virtual-machine: (next)
~/projects/libmesh/moose/test
$ mamba list 
# packages in environment at /home/pengwei/mambaforge3/envs/libmesh:
#
# Name                    Version                   Build  Channel\
```

**Reply by milljm** [reply]
It looks like you started with these instructions: https://mooseframework.inl.gov/getting_started/installation/conda.html

and then went with these: https://mooseframework.inl.gov/getting_started/installation/hpc_install_moose.html

Is this correct?

**Reply by PengWei97** [reply]
Yes, I used https://mooseframework.inl.gov/getting_started/installation/conda.html first, then created a folder `./projects/libmesh/moose` to clone the moose repository, and then based on https:/ /mooseframework.inl.gov/getting_started/installation/hpc_install_moose.html to install.

**Reply by milljm** [reply]
I think you followed one too many instructions, based on what I am seeing in the libmesh_diagnostic.log. Was there a reason to try and perform the HPC instructions? The Conda instructions are best used for single workstations/laptops (which I think you have).

The HPC instructions are for those wishing to use a high performance job scheduling capable environment. Where typically you need optimized binaries designed to work specifically with that system's MPI infrastructure.

If following the HPC instructions was an accident, I believe the only steps left for you to do is:

```bash
mamba activate moose
mamba install moose-libmesh moose-tools
mamba deactivate
mamba activate moose
cd ~/projects/libmesh/moose/test
make clobber all       # to clean up any previous build
make -j 6
./run_tests -j 6
```

**Reply by PengWei97** [reply]
> Was there a reason to try and perform the HPC instructions? 

My purpose is to be able to link libmesh directly, so that I can have a deeper understanding of moose, considering that I need to create my own model, this is related to mesh data.

I think my libmesh has been installed, but when I execute the following code,
```bash
cd ~/projects/moose/test
make -j4
```
It shows `make: python-config: Command not found`, how to solve this?

```shell
Compiling C++ (in opt mode) /home/pengwei/projects/moose/framework/build/unity_src/mesh_Unity.C...
Building and linking /home/pengwei/projects/moose/framework/contrib/hit/hit.so...
/home/pengwei/projects/moose/framework/contrib/hit/hit.cpp:4:10: fatal error: Python.h: No such file or directory
    4 | #include "Python.h"
      |          ^~~~~~~~~~
compilation terminated.
Compiling C++ (in opt mode) /home/pengwei/projects/moose/framework/contrib/exodiff/exodiff.C...
make: *** [/home/pengwei/projects/moose/framework/moose.mk:124: /home/pengwei/projects/moose/framework/contrib/hit/hit.so] Error 1
make: *** Waiting for unfinished jobs....
```

**Reply by PengWei97** [reply]
Woohoo!! It works using `sudo apt-get install python3-dev`
![image](https://user-images.githubusercontent.com/56183775/175561270-20c0de0b-7ac1-419f-9aad-f5de7ef40199.png)

**Reply by milljm** [reply]
🎉  Nice! Looks like you got it sorted!

FYI, you can still use Conda instead of the manual install (if you choose). You simply install only the things that `moose-libmesh` requires. Which I admit, is not documented. But that package install would look something like:

```bash
mamba create -n just-petsc moose-petsc
mamba activate just-petsc
cd moose/scripts
export METHODS=oprof MOOSE_JOBS=6   # build profile method, using 6 cores
./update_and_rebuild_libmesh.sh
export METHOD=oprof                 # build moose using same method as libMesh
cd ../test
make -j 6
./run_tests --oprof                 # run tests using the moose_test-oprof binary
```
There are several 'METHOD(S)' available, FYI; `oprof, opt, devel, dbg`
The description of what each means can be found here: https://mooseframework.inl.gov/moose/application_development/build_system.html
